x/incentive/keeper/unbonding.go

```diff
@@ -8,34 +8,117 @@ import (
 // addUnbonding creates an unbonding and adds it to the account's current unbondings in the store.
 // Assumes the validity of the unbonding has already been checked. Also updates unbonding amounts
 // indirectly by calling setUnbondings.
-func (k Keeper) addUnbonding(ctx sdk.Context, addr sdk.AccAddress, uToken sdk.Coin, tier incentive.BondTier) error {
-	if err := k.decreaseBond(ctx, addr, tier, uToken); err != nil {
+func (k Keeper) addUnbonding(ctx sdk.Context, addr sdk.AccAddress, uToken sdk.Coin) error {
+	if err := k.decreaseBond(ctx, addr, uToken); err != nil {
 		return err
 	}
+	unbondingDuration := k.GetParams(ctx).UnbondingDuration
+	if unbondingDuration == 0 {
+		// For unbonding duration zero, return after decreasing bonded amount
+		// without creating an unbonding struct
+		return nil
+	}
+	currentTime := k.GetLastRewardsTime(ctx)
 	unbonding := incentive.Unbonding{
-		Amount: uToken,
-		End:    k.getLastRewardsTime(ctx) + k.unbondTime(ctx, tier),
+		UToken: uToken,
+		// Start and end time are stored based on current parameters, and
+		// the stored end time does not change even if the module's unbonding
+		// duration parameter is changed. The unbonding will still end early
+		// if that parameter is reduced though.
+		Start: currentTime,
+		End:   currentTime + unbondingDuration,
 	}
 	unbondings := incentive.AccountUnbondings{
 		Account:    addr.String(),
-		Tier:       uint32(tier),
-		Denom:      uToken.Denom,
-		Unbondings: append(k.getUnbondings(ctx, addr, uToken.Denom, tier), unbonding),
+		UToken:     uToken.Denom,
+		Unbondings: append(k.getUnbondings(ctx, addr, uToken.Denom), unbonding),
 	}
 	return k.setUnbondings(ctx, unbondings)
 }
 
-// unbondTime returns how long a given tier must wait to unbond, in seconds.
-// returns zero on invalid unbond tier, though this will not happen in practice.
-func (k Keeper) unbondTime(ctx sdk.Context, tier incentive.BondTier) uint64 {
-	switch tier {
-	case incentive.BondTierLong:
-		return k.getUnbondingDurationLong(ctx)
-	case incentive.BondTierMiddle:
-		return k.getUnbondingDurationMiddle(ctx)
-	case incentive.BondTierShort:
-		return k.getUnbondingDurationShort(ctx)
-	default:
-		return 0
+// finishUnbondings finishes any unbondings on an account which have reached their end time
+func (k Keeper) finishUnbondings(ctx sdk.Context, addr sdk.AccAddress) error {
+	time := k.GetLastRewardsTime(ctx)
+	duration := k.GetParams(ctx).UnbondingDuration
+	bondedDenoms := k.getAllBondDenoms(ctx, addr)
+
+	for _, denom := range bondedDenoms {
+		storedUnbondings := k.getUnbondings(ctx, addr, denom)
+		newUnbondings := incentive.NewAccountUnbondings(
+			addr.String(),
+			denom,
+			[]incentive.Unbonding{},
+		)
+		for _, u := range storedUnbondings {
+			if u.End > time && u.Start+duration > time {
```

Comment:
why do we have `u.End` if we check `u.Start+duration` anyway? It seams that `u.Start+duration` was added to accomodate duration updates, while some unbondings are still active. In that case `u.End` is redundand and potentially false.

Reply:
There are two possible cases for param change here:
- Duration was increased. In this case, ongoing unbondings should not be extended so they need to remember their original end time.
- Duration was decreased. In this case, ongoing unbondings should be shortened so they need to use start + duration instead of original end.

To accomodate both scenarios (without iterating over all unbondings for all users on `SetParams` the current logic is used.